File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.java
import java.util.*;

class Solution {
    public ArrayList<Integer> solution(String today, String[] terms, String[] privacies) {
        ArrayList<Integer> answer = new ArrayList<>();
        
        HashMap<String, Integer> map = new HashMap<>();
        
        for(String term : terms){
            String[] temps = term.split(" ");
            int mon = Integer.parseInt(temps[1]);
            
            map.put(temps[0], mon);
        }
        
        String[] dates = today.split("\\.");
        int year = Integer.parseInt(dates[0]);
        int month = Integer.parseInt(dates[1]);
        int day = Integer.parseInt(dates[2]);
        
        int idx = 1;
        
        for(String privacy : privacies){
            String[] temps = privacy.split(" ");
            String[] date = temps[0].split("\\.");
            int duration = map.get(temps[1]);
            
            int y = Integer.parseInt(date[0]);
            int m = Integer.parseInt(date[1]);
            int d = Integer.parseInt(date[2]);
            
            m += (duration % 12);
            
            if(duration >= 12){
               y += (duration / 12); 
            }
            
            if(m > 12){
               y += (m / 12);
               m %= 12;
            }
            
            d--;
            
            if(d == 0){
                m -= 1;
                d = 28;
                
                if(m == 0){
                    y--;
                    m = 12;
                }
                
            }
            
            if(year > y){
                answer.add(idx);
            }else if((year == y) && (month > m)){
                answer.add(idx);
            }else if((year == y) && (month == m) && (day > d))
                answer.add(idx);
            
            idx++;
        }
         
        return answer;
    }
}

File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.py
def solution(today, terms, privacies):
    answer = []
    d2 = today.split(".")

    for i in range(len(privacies)):
        s = privacies[i].split(" ")
        d1 = s[0].split(".")
        for j in range(len(terms)):
            t = terms[j].split(" ")
            if s[1] == t[0]:
                d1[1] = int(d1[1]) + int(t[1]) % 12
                d1[2] = int(d1[2]) - 1
                if int(t[1]) >= 12:
                    d1[0] = int(d1[0]) + (int(t[1])//12)
                if d1[1] > 12:
                    d1[0] = int(d1[0]) + (d1[1]//12)
                    d1[1] = d1[1] - ((d1[1]//12) * 12)
                if d1[2] == 0:
                    d1[1] -= 1
                    d1[2] = 28
                    if d1[1] == 0:
                        d1[0] -= 1
                        d1[1] = 12
                break
        if int(d2[0]) == int(d1[0]):
            if int(d2[1]) > int(d1[1]):
                answer.append(i + 1)
            elif int(d2[1]) == int(d1[1]):
                if int(d2[2]) <= int(d1[2]):
                    continue
                else:
                    answer.append(i + 1)
            else:
                continue
        elif int(d2[0]) > int(d1[0]):
            answer.append(i + 1)
        else:
            continue
    return answer